File: src/test/scala/org/apache/flume/source/ftp/EmbeddedFtpSourceTest.scala
package org.apache.flume.source.ftp

import java.nio.file.{FileSystems, Path, Files}
import java.nio.file.attribute.PosixFilePermission
import javax.annotation.concurrent.NotThreadSafe

import com.typesafe.scalalogging.slf4j.LazyLogging
import org.apache.flume.channel.ChannelProcessor
import org.apache.flume.source.utils.FTPSourceEventListener
import org.apache.flume.{Event, PollableSource, Context}
import org.apache.flume.source.{FtpSourceCounter, FTPSource, TestFileUtils}
import org.apache.flume.source.ftp.server.EmbeddedFTPServer
import org.mockito.{MockitoAnnotations, Mock}
import org.mockito.Mockito._
import org.testng.Assert._
import org.testng.annotations.{Test, AfterMethod, BeforeMethod}
import scala.collection.JavaConversions._

/**
 * Basic integration tests for the Keedios' Flume FTP Source,
 *
 * Created by luca on 30/1/15.
 */
@NotThreadSafe
class EmbeddedFtpSourceTest extends EmbeddedFTPServer with TestFileUtils with LazyLogging with AbstractFtpSourceTest {

  logger.info("homeDir: "+EmbeddedFTPServer.homeDirectory.toFile.getAbsolutePath)

  @Test
  def testProcessNoFile() = {
    val proc = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc)
    assertEquals(ftpSourceCounter.getFilesCount, 0)
    assertEquals(ftpSourceCounter.getFilesProcCount,0)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)


  }

  @Test
  def testProcessNewFile(): Unit ={
    val tmpFile = createTmpFile(EmbeddedFTPServer.homeDirectory)

    val proc = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc)
    assertEquals(ftpSourceCounter.getFilesCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)
    cleanup(tmpFile)
  }

  @Test
  def testProcessNewFileInNewFolder(): Unit = {
    val tmpDir = createTmpDir(EmbeddedFTPServer.homeDirectory)

    val tmpFile = createTmpFile(tmpDir)

    val proc = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc)
    assertEquals(ftpSourceCounter.getFilesCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)

    cleanup(tmpDir)
  }

  /**
   * Creates a new empty file in the ftp root,
   * creates a new directory in the ftp root and an empty file inside of it.
   */
  @Test
  def testProcessMultipleFiles0(): Unit = {
    val tmpDir = createTmpDir(EmbeddedFTPServer.homeDirectory)

    val tmpFile0 = createTmpFile(EmbeddedFTPServer.homeDirectory)
    val tmpFile1 = createTmpFile(tmpDir)

    val proc = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc)
    assertEquals(ftpSourceCounter.getFilesCount,2)
    assertEquals(ftpSourceCounter.getFilesProcCount,2)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)

    cleanup(tmpFile0)
    cleanup(tmpDir)
  }

  /**
   * Tries to access a file without permissions
   */
  @Test
  def testProcessNoPermission(): Unit = {
    val tmpFile0 = createTmpFile(EmbeddedFTPServer.homeDirectory)

    Files.setPosixFilePermissions(tmpFile0,
      Set[PosixFilePermission](PosixFilePermission.OWNER_WRITE))

    val proc = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc)
    assertEquals(ftpSourceCounter.getFilesCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCount,0)
    assertEquals(ftpSourceCounter.getFilesProcCountError,1)

    cleanup(tmpFile0)
  }

  @Test
  def testProcessNotEmptyFile(): Unit = {
    val tmpFile0 = createTmpFile(EmbeddedFTPServer.homeDirectory)
    appendASCIIGarbageToFile(tmpFile0)
    val proc = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc)
    assertEquals(ftpSourceCounter.getFilesCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)

    val map = ftpSource.loadMap("hasmap.ser")

    logger.info(tmpFile0.toFile.getName)

    val filename = "//"+tmpFile0.toFile.getName

    assertEquals(map.get(filename), 81L*100L)

    cleanup(tmpFile0)
  }

  @Test
  def testProcessModifiedFile(): Unit = {
    val tmpFile0 = createTmpFile(EmbeddedFTPServer.homeDirectory)
    appendASCIIGarbageToFile(tmpFile0)
    val proc0 = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc0)
    assertEquals(ftpSourceCounter.getFilesCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)

    appendASCIIGarbageToFile(tmpFile0, 1000, 100)

    val proc1 = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc0)
    assertEquals(ftpSourceCounter.getFilesCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCount,1)
    assertEquals(ftpSourceCounter.getFilesProcCountError,0)

    val map = ftpSource.loadMap("hasmap.ser")

    logger.info(tmpFile0.toFile.getName)

    val filename = "//"+tmpFile0.toFile.getName

    assertEquals(map.get(filename), 81L*100L + 1000L*101L)

    cleanup(tmpFile0)
  }

  /**
   * Creates N temporary non-empty files in the
   * FTP root dir and process it using the FTP source.
   */
  @Test
  def testProcessMultipleFiles1(): Unit ={
    val totFiles = 100

    val files = for (i <- 1 to totFiles) yield {
      val tmpFile0 = createTmpFile(EmbeddedFTPServer.homeDirectory)
      appendASCIIGarbageToFile(tmpFile0)

      if (i == 8){
        Files.setPosixFilePermissions(tmpFile0,
          Set[PosixFilePermission](PosixFilePermission.OWNER_WRITE))
      }

      tmpFile0
    }

    val proc0 = ftpSource.process
    assertEquals(PollableSource.Status.READY, proc0)
    assertEquals(ftpSourceCounter.getFilesCount,totFiles)
    assertEquals(ftpSourceCounter.getFilesProcCount,totFiles-1)
    assertEquals(ftpSourceCounter.getFilesProcCountError,1)

    cleanup(files:_*)
  }


}

File: src/test/java/org/apache/flume/source/ftp/EmbeddedFtpSourceTest.java
package org.apache.flume.source.ftp;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Files;
import java.nio.file.attribute.PosixFilePermission;
import java.util.*;
import javax.annotation.concurrent.NotThreadSafe;

import org.apache.flume.EventDeliveryException;
import org.apache.flume.PollableSource;
import org.apache.flume.source.TestFileUtils;
import org.apache.flume.source.ftp.server.EmbeddedFTPServer;

import org.apache.flume.source.utils.FTPSourceEventListener;
import org.apache.ftpserver.ftplet.FtpException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testng.Assert;
import org.testng.annotations.AfterSuite;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;

/**
 * Basic integration tests for the Keedios' Flume FTP Source,
 *
 */
@NotThreadSafe
public class EmbeddedFtpSourceTest extends AbstractFtpSourceTest {

    private static Logger logger = LoggerFactory.getLogger(EmbeddedFtpSourceTest.class);

    static {
        logger.info("homeDir: " + EmbeddedFTPServer.homeDirectory.toFile().getAbsolutePath());
    }

    @Test
    public void testProcessNoFile() {
        try {
            PollableSource.Status proc = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 0);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 0);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);
        } catch (EventDeliveryException e) {
            Assert.fail();
        }
    }

    @Test(dependsOnMethods = "testProcessNoFile")
    public void testProcessNewFile() {
        Path tmpFile = null;
        try {
            tmpFile = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
            PollableSource.Status proc = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);
        } catch (IOException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup( tmpFile );
        }
    }

    @Test(dependsOnMethods = "testProcessNewFile")
    public void testProcessNewFileInNewFolder() {
        Path tmpDir = null;
        Path tmpFile = null;
        try {
            tmpDir = TestFileUtils.createTmpDir(EmbeddedFTPServer.homeDirectory);
            tmpFile = TestFileUtils.createTmpFile(tmpDir);

            PollableSource.Status proc = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);
        } catch (IOException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup(Arrays.asList(tmpFile, tmpDir));
        }

    }

    /**
     * Creates a new empty file in the ftp root,
     * creates a new directory in the ftp root and an empty file inside of it.
     */
    @Test(dependsOnMethods = "testProcessNewFileInNewFolder")
    public void testProcessMultipleFiles0() {
        Path tmpDir = null;
        Path tmpFile0 = null;
        Path tmpFile1 = null;
        try {
            tmpDir = TestFileUtils.createTmpDir(EmbeddedFTPServer.homeDirectory);
            tmpFile0 = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
            tmpFile1 = TestFileUtils.createTmpFile(tmpDir);

            PollableSource.Status proc = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 2);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 2);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);
        } catch (IOException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup(Arrays.asList(tmpFile0, tmpFile1, tmpDir));
        }
    }

    /**
     * Tries to access a file without permissions
     */
    @Test(dependsOnMethods = "testProcessMultipleFiles0")
    public void testProcessNoPermission() {
        Path tmpFile0 = null;
        try {
            tmpFile0 = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
            Files.setPosixFilePermissions(tmpFile0,
                    new HashSet<PosixFilePermission>(Arrays.asList(PosixFilePermission.OWNER_WRITE)));

            PollableSource.Status proc = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 0);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 1);
        } catch (IOException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup(tmpFile0);
        }

    }

    @Test(dependsOnMethods = "testProcessNoPermission")
    public void testProcessNotEmptyFile() {
        Path tmpFile0 = null;
        try {
            tmpFile0 = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
            TestFileUtils.appendASCIIGarbageToFile(tmpFile0);
            PollableSource.Status proc = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);

            HashMap<String, Long> map = ftpSource.loadMap("hasmap.ser");

            String filename = "//"+tmpFile0.toFile().getName();

            Assert.assertEquals( Long.valueOf(map.get(filename)), Long.valueOf(81L * 100L));
        } catch (IOException|ClassNotFoundException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup(tmpFile0);
        }
    }

    @Test(dependsOnMethods = "testProcessNotEmptyFile")
    public void testProcessModifiedFile() {
        Path tmpFile0 = null;
        try {
            tmpFile0 = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
            TestFileUtils.appendASCIIGarbageToFile(tmpFile0);
            PollableSource.Status proc0 = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc0);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);

            TestFileUtils.appendASCIIGarbageToFile(tmpFile0, 1000, 100);

            PollableSource.Status proc1 = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc0);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 0);

            Map<String,Long> map = ftpSource.loadMap("hasmap.ser");

            String filename = "//"+tmpFile0.toFile().getName();

            Assert.assertEquals(Long.valueOf(map.get(filename)), Long.valueOf(81L * 100L + 1000L * 101L));

        } catch (IOException|EventDeliveryException|ClassNotFoundException e) {
            Assert.fail();
        } finally {
            cleanup(tmpFile0);
        }
    }

    /**
     * Creates N temporary non-empty files in the
     * FTP root dir and process it using the FTP source.
     */
    @Test(dependsOnMethods = "testProcessModifiedFile")
    public void testProcessMultipleFiles1() {
        int totFiles = 100;
        List<Path> files = new LinkedList<>();

        try {
            for (int i = 1; i <= totFiles; i++) {
                Path tmpFile0 = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
                TestFileUtils.appendASCIIGarbageToFile(tmpFile0);

                if (i == 8) {
                    Files.setPosixFilePermissions(tmpFile0,
                            new HashSet<>(Arrays.asList(PosixFilePermission.OWNER_WRITE)));
                }

                files.add(tmpFile0);
            }

            PollableSource.Status proc0 = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc0);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), totFiles);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), totFiles - 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 1);
        } catch (IOException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup(files);
        }
    }

    @Test(dependsOnMethods = "testProcessMultipleFiles1")
    public void testFtpFailure() {
        class MyEventListener extends FTPSourceEventListener {
            @Override
            public void fileStreamRetrieved()  {
                logger.info("Stopping server");
                EmbeddedFTPServer.ftpServer.stop();
            }
        }
        ftpSource.setListener(new MyEventListener());

        String[] directories = EmbeddedFTPServer.homeDirectory.toFile().list();

        logger.info("Found files: ");

        for (String directory : directories) {
            logger.info(directory);
        }

        Path tmpFile0 = null;
        try {
            tmpFile0 = TestFileUtils.createTmpFile(EmbeddedFTPServer.homeDirectory);
            TestFileUtils.appendASCIIGarbageToFile(tmpFile0, 1000, 100);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 0);

            PollableSource.Status proc0 = ftpSource.process();
            Assert.assertEquals(PollableSource.Status.READY, proc0);
            Assert.assertEquals(ftpSourceCounter.getFilesCount(), 1);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCount(), 0);
            Assert.assertEquals(ftpSourceCounter.getFilesProcCountError(), 1);
        } catch (IOException|EventDeliveryException e) {
            Assert.fail();
        } finally {
            cleanup(tmpFile0);
        }
    }
}
